backend/pkg/rate/limit_http.go

```diff
@@ -0,0 +1,204 @@
+		limiter:       limiter,
+		eventTemplate: t,
+	}
+	return nil
+}
+
+func (h *HTTPLimiter) MatchHTTPPattern(
+	pattern, groupTemplate string,
+) error {
+	var (
+		t   *template.Template
+		err error
+	)
+	if groupTemplate != "" {
+		// Compile eventTemplate:
+		t, err = h.template.Clone()
+		if err == nil {
+			_, err = t.Parse(groupTemplate)
+		}
+		if err != nil {
+			return fmt.Errorf("error parsing group_template: %w", err)
+		}
+	}
+	limiterHandle := handle{
+		HTTPLimiter:   h,
+		groupTemplate: t,
+	}
+	h.mux.Handle(pattern, limiterHandle)
+	return nil
+}
+
+type handle struct {
+	*HTTPLimiter
+	groupTemplate *template.Template
+}
+
+// ServeHTTP implements a basic http.Handler so that handler can be used
+// as a handler for the mux. It will only write on errors and is expected
+// to continue to the actual handler on success.
+func (h *HTTPLimiter) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	if h.rewriteRequests {
+		r = rest.RewriteForwardedRequest(r)
+	}
+	res, err := h.Reserve(r)
+	if err != nil {
+		hdr := w.Header()
+		hdr.Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusInternalServerError)
+		_ = json.NewEncoder(w).Encode(rest.Error{
+			Err:       "internal server error",
+			RequestID: requestid.FromContext(ctx),
+		})
+		return
+	}
+	if res == nil || res.OK() {
+		return
+	} else {
+		hdr := w.Header()
+		hdr.Set("Content-Type", "application/json")
+		retryAfter := math.Ceil(res.Delay().Abs().Seconds())
+		hdr.Set("Retry-After", fmt.Sprint(retryAfter))
+		w.WriteHeader(http.StatusTooManyRequests)
+		_ = json.NewEncoder(w).Encode(rest.Error{
+			Err:       ErrTooManyRequests.Error(),
+			RequestID: requestid.FromContext(ctx),
+		})
+	}
+}
+
+func (h *HTTPLimiter) MiddlewareGin(c *gin.Context) {
+	r := c.Request
+	if h.rewriteRequests {
+		r = rest.RewriteForwardedRequest(r)
+	}
+	res, err := h.Reserve(r)
+	if err != nil {
+		rest.RenderInternalError(c, err)
+		c.Abort()
+	}
+	if res == nil || res.OK() {
+		c.Next()
+	} else {
+		retryAfter := math.Ceil(res.Delay().Abs().Seconds())
+		c.Header("Retry-After", fmt.Sprint(retryAfter))
+		rest.RenderError(c, http.StatusTooManyRequests, ErrTooManyRequests)
+		c.Abort()
+	}
+}
+
+func (m *HTTPLimiter) Reserve(r *http.Request) (Reservation, error) {
+	var b bytes.Buffer
+	limiter := m.DefaultGroup
+	templateData := m.templateData(r)
+	ctx := r.Context()
+	h, _ := m.mux.Handler(r)
+	hh, ok := h.(handle)
+	if ok {
+		if hh.groupTemplate != nil {
+			err := hh.groupTemplate.Execute(&b, templateData)
+			if err != nil {
+				return nil, fmt.Errorf("error executing ratelimit group template: %w", err)
+			}
+			group := m.limiterGroups[b.String()]
+			if group != nil {
+				limiter = group
+			}
+			b.Reset()
+		}
+	}
+	err := limiter.eventTemplate.Execute(&b, templateData)
+	if err != nil {
+		return nil, fmt.Errorf("error executing template for event ID: %w", err)
+	}
+	if limiter == nil {
+		return nil, nil
```

Comment:
I am not a fan of nil,nil in calls like Reserve are you sure this is what we want? btw we already dereferenced the limiter a couple of lines above.

Reply:
This was a leftover, I will remove it. I believe this is a constant condition since it should always be set. :memo: